auction-house/program/src/receipt/mod.rs

```diff
@@ -552,23 +563,29 @@ pub fn print_purchase_receipt<'info>(
 
     purchase.try_serialize(&mut *purchase_receipt_account.try_borrow_mut_data()?)?;
 
-    let mut listing_receipt_data = listing_receipt_info.try_borrow_mut_data()?;
-    let mut listing_receipt_data_slice: &[u8] = &listing_receipt_data;
-
-    let mut listing_receipt = ListingReceipt::try_deserialize(&mut listing_receipt_data_slice)?;
+    // since partial buy now exists, need to check token_size before closing listing receipt account
 
-    listing_receipt.purchase_receipt = Some(purchase_receipt_account.key());
+    let token_account_data = SplAccount::unpack(&token_account.data.borrow())?;
 
-    listing_receipt.try_serialize(&mut *listing_receipt_data)?;
+    if token_account_data.amount == 0 {
+        let mut listing_receipt_data = listing_receipt_info.try_borrow_mut_data()?;
+        let curr_listing_receipt_lamp = listing_receipt_info.lamports();
+        **listing_receipt_info.lamports.borrow_mut() = 0;
+        sol_memset(*listing_receipt_data, 0, LISTING_RECEIPT_SIZE);
+        **bookkeeper.lamports.borrow_mut() = bookkeeper
+            .lamports()
+            .checked_add(curr_listing_receipt_lamp)
+            .ok_or(AuctionHouseError::NumericalOverflow)?;
+    }
 
     let mut bid_receipt_data = bid_receipt_account.try_borrow_mut_data()?;
-    let mut bid_receipt_slice: &[u8] = &bid_receipt_data;
-
-    let mut bid_receipt = BidReceipt::try_deserialize(&mut bid_receipt_slice)?;
-
-    bid_receipt.purchase_receipt = Some(purchase_receipt_account.key());
-
-    bid_receipt.try_serialize(&mut *bid_receipt_data)?;
+    let curr_bid_receipt_lamp = bid_receipt_info.lamports();
+    **bid_receipt_info.lamports.borrow_mut() = 0;
+    sol_memset(*bid_receipt_data, 0, BID_RECEIPT_SIZE);
+    **bookkeeper.lamports.borrow_mut() = bookkeeper
+        .lamports()
+        .checked_add(curr_bid_receipt_lamp)
+        .ok_or(AuctionHouseError::NumericalOverflow)?;
 
```

Comment:
Closing the listing and bid receipt when execute sale is fired messes up indexing of these accounts with geyser since when the account is passed to the geyser on account update the data will be empty. The indexer won't be able to correctly process the account. It has the address but the discriminator is cleared as well.

Close receipt should be a separate permissionless instruction that sends funds to the bookkeeper (as it is now) but only let the instruction run if canceled_at or purchase_receipt written to the bid receipt or listing receipt.

This would allow indexers to process the account change and then at a later time close our the accounts after they've been indexed properly. 

At Holaplex, we are speccing v2 of our indexer and will likely track address to accounts type so when data is empty can still lookup the table associated with the account by its address.

Reply:
Holaplex accounts for cancel offer, cancel listings, and execute sale instructions now so close away.

https://github.com/holaplex/indexer/tree/dev/crates/indexer/src/geyser/instructions